### ISSUE TITLE
Better Wiki support

### ISSUE THREAD

**Issue by zeigerpuppy** [codimd/server#90]
CodiMD already has the underlying features necessary for using it as a wiki (especially arbitrary named page creation).  There are a couple of minor changes that would make this much easier to use, however:

[ ] ability to add internal links with the `[[NewPage]]` syntax (currently, a user needs to use `[NewPage](/NewPage)`
[ ] A way of seeing backlinks (Maybe a left sided TOC like menu on the view mode only would be an option)
[ ] An editor button for adding wiki links

**Comment by ccoenen** [comment]
I am against using mediawiki-style syntax just to look more like mediawiki. I would like to avoid deviating from the markdown standard for things that are already part of the markdown standard (i.e. `[[links]]` )

I am not at all against points 2 and 3! Those would make using CodiMD more pleasant!

**Comment by zeigerpuppy** [comment]
I can understand the concern of deviating from standard markdown.  Incidentally, the syntax is also used in the [Gollum wiki](https://github.com/gollum/gollum/wiki#tags).  For internal development/documentation, our team currently uses a customised Gollum instance with PlantUML support but I could easily see this replaced by CodiMD as the real-time editor is much more user friendly.

**Comment by almereyda** [comment]
Yes, wiki markup, and if only the double brackets, are a quite widely adopted pattern for internal links. We would only have to consider how we turn page titles enclosed within those into viable slugs that render a valid URL.

**Comment by SISheogorath** [comment]
How about we add a simple internal link style like `codimd://<noteid>` which could be used like `[title](codimd://<noteid>)` this would provide proper support for the linking and since it's a proper URL it can be easily replaced on exports, etc. With a real http link.

**Comment by bitinerant** [comment]
You can already do something like this in CodiMD: `[title of other page](/neYk5utpQH6cxukGVUcHrQ)`. I *think* the goal of the new syntax is to not have to specify the title in the actual markdown, i.e. that CodiMD would fetch it from the target page. Maybe I misunderstand.

**Comment by zeigerpuppy** [comment]
The main advantage of having a simple wiki-style link is in the act of writing a wiki.
The author can make new pages based on concept rather than having to worry about whether that page exists or needs to be created.  This also has the advantage of supporting incremental creation, where the author can have the happy coincidence of referring to a wiki page and finding it already partly or fully created.  Therefore, a wiki page should be referred to by one tag only, which ideally aligns with it's subject matter.  
I can understand @ccoenen's concern regarding syntax.  Nevertheless, to be useful for a wiki, the syntax should require minimal fore-planning on the part of the author.  An alternative syntax that would be relatively easy to use could be:
- `[<Wiki Page>]`

As a side note, it's worth considering redundancy titles too.  For instance, its a good thing for a wiki to ignore capitalisation, `Wiki Page` *should* equal `wiki page`.  It's also important to handle spaces and punctuation, either by stripping them or having specific rules for page creation.
In the Gollum wiki, spaces are parsed to `-` and full stops to `dot`.  This is probably a good idea as it allows the user to have some freedom in creating page titles which still parse properly as URLs.

> the reason I keep referring to the Gollum wiki is because it supports markdown files, so has largely dealt with some of the issues likely to be encountered.

**Comment by almereyda** [comment]
Yes, slug generation from those internal links can be challenging, esp. when transforming Unicode into ASCII for a URL respectively pad link. This is how @fedwiki approaches the subject:

- https://github.com/fedwiki/wiki-client/blob/3e0a1dd499b5eb637eeb4a03970c9f0a4310e619/lib/page.coffee#L17-L18
- http://ward.asia.wiki.org/view/beyond-slugs

**Comment by SISheogorath** [comment]
I think the main reason, why this doesn't make a lot of sense for us is that in default, we don't run in free-url mode. Therefore we don't really use slugs. And that's why I would favor a non-slug-based solution for this problem as it would solve it for everyone who wants to reference documents, not just the people who use free-url or maybe some special future Wiki-mode.

**Comment by wscheicher** [comment]
Just an idea: Since (if i understand correctly) the main conflict here is that is easier to write `[[A]]` than `[A](/A)` on one hand, and the standard on the other, could that be solved by some sort of autoreplace/autocomplete feature of the editor?
